Enable compliant manylinux1 builds on Azure

pskopnik opened the pull request:
This PR (hopefully) enables manylinux1 builds by using the [konstin2/maturin:master][konstin2-maturin] Docker image as outlined in https://github.com/ijl/orjson/issues/2#issuecomment-589150572.

I've completely removed Python 3.9 for now, as it is not yet included in the manylinux Docker image. The simplest solution would be to fallback to the previous template for Python 3.9.

[konstin2-maturin]: https://hub.docker.com/r/konstin2/maturin

ijl commented:
I would be okay using the manylinux1 containers directly in place of the python containers, but don't see a reason to make a change when the current build process installs rust, maturin, etc. Could you modify the `path` parameter to be appropriate for the `/opt/bin/<target>`? Can you run benchmarks on both an existing release and the same release built in a manylinux1 wheel and confirm no significant change?

pskopnik commented:
## Benchmark results

`pypi` is v2.5.0 as published on PyPi. `ml` is the manylinux1 build of the v2.5.0 tag, built by extracting the commands from `azure-linux-container.yml` into a bash script and running it in `quay.io/pypa/manylinux1_x86_64`. All tests pass.

The benchmarks are using Python 3.7. I downgraded to rustc 1.43.0-nightly (e620d0f33 2020-02-18) due to hsivonen/encoding_rs#49.

### Latency

There seems to be no significant regression.

<details>

<summary>Brief Comparison</summary>

```
------------------------------------------------------------------------------ benchmark 'canada.json deserialization': 10 tests ------------------------------------------------------------------------------
Name (time in ms)                                   Min                Max               Mean            StdDev             Median               IQR            Outliers      OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_loads_canada_orjson (pybench_pypi)         21.9589 (1.69)     22.6810 (1.66)     22.2522 (1.67)     0.3018 (1.14)     22.1318 (1.67)     0.4805 (1.37)          1;0  44.9393 (0.60)          5         100
test_loads_canada_orjson (pybench_ml)           22.4276 (1.73)     23.7840 (1.74)     22.8600 (1.72)     0.5381 (2.02)     22.6594 (1.71)     0.5369 (1.53)          1;0  43.7446 (0.58)          5         100
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

--------------------------------------------------------------------------------- benchmark 'canada.json serialization': 10 tests ---------------------------------------------------------------------------------
Name (time in ms)                                   Min                 Max                Mean            StdDev              Median               IQR            Outliers       OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_dumps_canada_orjson (pybench_ml)            7.3742 (1.0)        7.4050 (1.0)        7.3836 (1.0)      0.0126 (1.0)        7.3820 (1.0)      0.0134 (1.0)           1;0  135.4350 (1.0)           5         136
test_dumps_canada_orjson (pybench_pypi)          7.4213 (1.01)       7.8827 (1.06)       7.6122 (1.03)     0.2458 (19.56)      7.4527 (1.01)     0.4563 (34.09)         2;0  131.3672 (0.97)          5         135
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------------------------------ benchmark 'citm_catalog.json deserialization': 10 tests -------------------------------------------------------------------------------
Name (time in ms)                                         Min                Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_loads_citm_catalog_orjson (pybench_pypi)          5.5614 (1.0)       5.9377 (1.04)      5.7771 (1.01)     0.1566 (104.75)    5.7761 (1.01)     0.2591 (154.05)        2;0  173.0967 (0.99)          5         181
test_loads_citm_catalog_orjson (pybench_ml)            5.6999 (1.02)      5.7339 (1.0)       5.7138 (1.0)      0.0164 (10.95)     5.7051 (1.0)      0.0296 (17.57)         1;0  175.0148 (1.0)           5         176
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------------------------------- benchmark 'citm_catalog.json serialization': 10 tests --------------------------------------------------------------------------------
Name (time in ms)                                         Min                Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_dumps_citm_catalog_orjson (pybench_pypi)          1.2544 (1.0)       1.2851 (1.0)       1.2650 (1.0)      0.0126 (1.0)       1.2582 (1.0)      0.0164 (1.0)           1;0  790.5002 (1.0)           5        1000
test_dumps_citm_catalog_orjson (pybench_ml)            1.2997 (1.04)      1.3694 (1.07)      1.3281 (1.05)     0.0345 (2.74)      1.3053 (1.04)     0.0609 (3.71)          1;0  752.9537 (0.95)          5        1000
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------- benchmark 'github.json deserialization': 10 tests -----------------------------------------------------------------------------------
Name (time in us)                                    Min                 Max                Mean            StdDev              Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_loads_github_orjson (pybench_ml)           244.4755 (1.0)      249.5816 (1.0)      245.5199 (1.0)      2.2708 (1.52)     244.5066 (1.0)      1.3372 (1.0)           1;1        4.0730 (1.0)           5       10000
test_loads_github_orjson (pybench_pypi)         246.9495 (1.01)     253.8624 (1.02)     249.7294 (1.02)     3.3100 (2.22)     248.1295 (1.01)     6.0267 (4.51)          1;0        4.0043 (0.98)          5       10000
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------- benchmark 'github.json serialization': 10 tests ------------------------------------------------------------------------------------
Name (time in us)                                    Min                 Max                Mean            StdDev              Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_dumps_github_orjson (pybench_pypi)          79.1468 (1.0)       81.4963 (1.0)       79.7666 (1.0)      0.9737 (6.35)      79.4028 (1.0)      0.6686 (5.90)          1;1       12.5366 (1.0)           5       12662
test_dumps_github_orjson (pybench_ml)            84.2483 (1.06)      84.6182 (1.04)      84.3462 (1.06)     0.1533 (1.0)       84.2918 (1.06)     0.1133 (1.0)           1;1       11.8559 (0.95)          5       11832
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------- benchmark 'twitter.json deserialization': 10 tests ----------------------------------------------------------------------------
Name (time in ms)                                   Min               Max              Mean            StdDev            Median               IQR            Outliers       OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_loads_twitter_orjson (pybench_pypi)         2.8461 (1.0)      2.9743 (1.0)      2.8992 (1.0)      0.0575 (2.51)     2.8779 (1.0)      0.1016 (2.79)          1;0  344.9176 (1.0)           5        1000
test_loads_twitter_orjson (pybench_ml)           2.8971 (1.02)     2.9874 (1.00)     2.9322 (1.01)     0.0388 (1.69)     2.9259 (1.02)     0.0650 (1.78)          1;0  341.0369 (0.99)          5        1000
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------------- benchmark 'twitter.json serialization': 10 tests ---------------------------------------------------------------------------------------
Name (time in us)                                       Min                   Max                  Mean             StdDev                Median                IQR            Outliers         OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_dumps_twitter_orjson (pybench_pypi)           765.8723 (1.0)        792.2828 (1.0)        775.6478 (1.0)      13.3399 (7.56)       766.1493 (1.0)      23.2326 (12.78)         1;0  1,289.2450 (1.0)           5        1306
test_dumps_twitter_orjson (pybench_ml)             811.4269 (1.06)       815.9349 (1.03)       812.9609 (1.05)      1.7639 (1.0)        812.1846 (1.06)      1.8174 (1.0)           1;0  1,230.0715 (0.95)          5        1230
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

</details>

[pybench_pypi.json.txt](https://github.com/ijl/orjson/files/4236413/pybench_pypi.json.txt)
[pybench_ml.json.txt](https://github.com/ijl/orjson/files/4236415/pybench_ml.json.txt)

### Memory

There seems to be no significant regression.

<details>

<summary><code>pypi</code> Results</summary>

#### twitter.json

| Library    |   import, read() RSS (MiB) |   loads() increase in RSS (MiB) |
|------------|----------------------------|---------------------------------|
| orjson     |                       29.6 |                             3.1 |
| ujson      |                       13   |                             4.6 |
| rapidjson  |                       14.2 |                             7.1 |
| simplejson |                       13.3 |                             2.6 |
| json       |                       12.9 |                             2.5 |

#### github.json

| Library    |   import, read() RSS (MiB) |   loads() increase in RSS (MiB) |
|------------|----------------------------|---------------------------------|
| orjson     |                       29   |                             0.3 |
| ujson      |                       12.4 |                             0.7 |
| rapidjson  |                       13.5 |                             0.9 |
| simplejson |                       12.7 |                             0.3 |
| json       |                       12.2 |                             0.4 |

#### citm_catalog.json

| Library    |   import, read() RSS (MiB) |   loads() increase in RSS (MiB) |
|------------|----------------------------|---------------------------------|
| orjson     |                       30.7 |                             8.5 |
| ujson      |                       14.1 |                            12   |
| rapidjson  |                       15.2 |                            35.5 |
| simplejson |                       14.5 |                            31.6 |
| json       |                       13.8 |                            25   |

#### canada.json

| Library    | import, read() RSS (MiB)   | loads() increase in RSS (MiB)   |
|------------|----------------------------|---------------------------------|
| orjson     | 31.2                       | 18.5                            |
| ujson      |                            |                                 |
| rapidjson  | 17.8                       | 19.9                            |
| simplejson | 17.1                       | 21.4                            |
| json       | 16.5                       | 21.4                            |


</details>

<details>

<summary><code>ml</code> Results</summary>

#### twitter.json

| Library    |   import, read() RSS (MiB) |   loads() increase in RSS (MiB) |
|------------|----------------------------|---------------------------------|
| orjson     |                       29.6 |                             2.8 |
| ujson      |                       13.1 |                             4.6 |
| rapidjson  |                       14.2 |                             7   |
| simplejson |                       13.3 |                             2.6 |
| json       |                       13   |                             2.5 |

#### github.json

| Library    |   import, read() RSS (MiB) |   loads() increase in RSS (MiB) |
|------------|----------------------------|---------------------------------|
| orjson     |                       29.1 |                             0.3 |
| ujson      |                       12.4 |                             0.7 |
| rapidjson  |                       13.6 |                             0.9 |
| simplejson |                       12.6 |                             0.3 |
| json       |                       12.1 |                             0.4 |

#### citm_catalog.json

| Library    |   import, read() RSS (MiB) |   loads() increase in RSS (MiB) |
|------------|----------------------------|---------------------------------|
| orjson     |                       30.7 |                             8.5 |
| ujson      |                       14.1 |                            12   |
| rapidjson  |                       15.2 |                            35.5 |
| simplejson |                       14.3 |                            31.4 |
| json       |                       13.9 |                            25   |

#### canada.json

| Library    | import, read() RSS (MiB)   | loads() increase in RSS (MiB)   |
|------------|----------------------------|---------------------------------|
| orjson     | 31.0                       | 18.5                            |
| ujson      |                            |                                 |
| rapidjson  | 17.8                       | 19.9                            |
| simplejson | 16.9                       | 21.4                            |
| json       | 16.5                       | 21.4                            |



</details>

pskopnik commented:
I had changed the template to get started on removing image dependent specifics from azure-pipelines.yml.
The maturin image would allow to factor out some dependencies.

Now, I reverted these changes. I've added $HOME to the $PATH components, I hope bash can expand it properly.

pip was issuing a warning due to the wrapper script. `$(interpreter) -m pip` ensures that the proper version of pip is used when multiple versions of Python (and pip) are installed.

What do you want to do about Python 3.9?

ijl commented:
Ok. python3.9 needs to remain, so that can contine using the non-manylinux wheel.

ijl commented:
https://github.com/pypa/manylinux/issues/482

pskopnik commented:
Checkout the latest version @ijl.

I've added an `auditwheel repair` check.
Reinstalling auditwheel is unfortunately necessary to downgrade to a compatible version of wheel (alternative: install all other dependencies in venv). The auditwheel created wheel is ignored, but if auditwheel cannot repair a wheel it exits with code != 0.

ijl commented on ci/azure-pipelines.yml:
```diff
@@ -43,11 +44,11 @@ jobs:
     interpreter: python3.7
     manylinux: 1
-    path: /home/vsts_azpcontainer/.local/bin:/home/vsts_azpcontainer/.cargo/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+    path: $HOME/.local/bin:$HOME/.cargo/bin:/opt/python/cp37-cp37m/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```
I tried using `$HOME` originally but it didn't expand when used as a variable. Do you not need `.local/bin` for the `--user` install?

pskopnik replied:
What do you mean with "didn't expand"? Azure Pipelines creates a bash script from the code. Is $HOME not set? I was unable to find the `/home/vsts_azpcontainer` value in the documentation (quick search). If $HOME is not set, maybe `~` will do the trick.

Yeah, `.local/bin` is required only due to `--user`. Dependencies could be installed globally.

ijl commented on ci/azure-pipelines.yml:
```diff
@@ -43,11 +44,11 @@ jobs:
   pool:
     vmImage: ubuntu-18.04
-  container: python:3.7-buster
+  container: quay.io/pypa/manylinux1_x86_64
```
This needs `:latest` at the end.

ijl commented on ci/azure-linux-container.yml:
```diff
@@ -3,25 +3,29 @@ parameters:
 - bash: PATH=$(path) RUSTFLAGS="-Zmutable-noalias -C target-feature=+sse2" maturin build --release --strip --manylinux $(manylinux) --interpreter $(interpreter) --target $(target)
   displayName: build
-- bash: PATH=$(path) pip install --user target/wheels/orjson*.whl
+- bash: PATH=$(path) $(interpreter) -m pip install --user --upgrade auditwheel && PATH=$(path) $(interpreter) -m auditwheel repair target/wheels/orjson*.whl
```
I've removed the install of wheel so this could now be installed with pip and the other build tools.

pskopnik replied:
`wheel` is already installed in all environments. `auditwheel` is available globally (`/usr/local/bin/`) via Python 3.7, so it doesn't have to be installed. The `pip install --upgrade ... wheel` leads to wheel being upgraded to a version not compatible with `auditwheel` atm. This is only relevant for the Python 3.7 build. The later install of auditwheel ensures that `wheel` is downgraded to the appropriate version.

Idk an idiomatic solution for this. So your solution seems fine, personally, I would install auditwheel directly before the `auditwheel` command in order to avoid any other installs messing with dependencies. Although both awkwardly overwrite the version shipped with manylinux. IMO, auditwheel should be installed to a venv in the manylinux images.

ijl commented:
/azp run

ijl commented:
I rebased and fixed the above and some other minor things in https://github.com/ijl/orjson/commit/ed51bac725a557ae106463d2478999159411492c. 2.5.1 will include the fix. Thank you.

pskopnik commented:
Thanks for merging and publishing :)

I've added some remarks to your review.

There is one last remaining call to pip, should be `$(interpreter) -m pip install ...`:
https://github.com/ijl/orjson/blob/ed51bac725a557ae106463d2478999159411492c/ci/azure-linux-container.yml#L14
